src/ship/GunManager.h

```diff
@@ -0,0 +1,222 @@
+	void UnmountWeapon(StringName hardpoint);
+	// Check if a weapon is attached to a specific mount
+	bool IsWeaponMounted(StringName hardpoint) const;
+
+	const std::vector<WeaponState> &GetWeapons() const { return m_weapons; }
+
+	uint32_t GetNumWeapons() const { return m_weapons.size(); }
+	const WeaponState *GetWeaponState(uint32_t numWeapon) const { return m_weapons.size() > numWeapon ? &m_weapons[numWeapon] : nullptr; }
+	uint32_t GetWeaponIndexForHardpoint(StringName hardpoint) const;
+
+	// ==========================================
+
+	// For AI/NPC ship usage, combines all weapons into a single group
+	void SetupDefaultGroups();
+	void SetTrackingTarget(const Body *target) { return SetGroupTarget(0, target); }
+	void SetAllGroupsFiring(bool firing = false);
+
+	// For player weapon management
+
+	// Assign a specific weapon to a firing group
+	// Weapons are by default all assigned to group 0
+	void AssignWeaponToGroup(uint32_t numWeapon, uint32_t group);
+	// Remove a weapon group in its entirety
+	void RemoveGroup(uint32_t group);
+
+	const std::vector<GroupState> &GetWeaponGroups() const { return m_groups; }
+
+	const Body *GetGroupTarget(uint32_t group);
+	void SetGroupTarget(uint32_t group, const Body *target);
+	void SetGroupFiring(uint32_t group, bool firing);
+	// Sets whether the group will authorize individual weapons to fire without a target within the gimbal arc of the weapon
+	void SetGroupFireWithoutTargeting(uint32_t group, bool enabled);
+
+	// TEMP: return the average lead position of a firing group - should query individual weapons instead
+	vector3d GetGroupLeadPos(uint32_t group);
+	// TEMP: return the average temperature state of a weapon group in 0..1
+	float GetGroupTemperatureState(uint32_t group);
+
+	// ==========================================
+
+	bool IsFiring() const { return m_isAnyFiring; }
+	bool IsGroupFiring(uint32_t group) const { return m_groups.size() > group ? m_groups[group].firing : false; }
+
+	// TODO: separate this functionality to a ship-wide cooling system
+	void SetCoolingBoost(float boost) { m_coolingBoost = boost; }
+
+	// Semi-hacky, would like to replace with a better eventing system of some sort
+	WeaponIndexSet GetGunsFiredThisFrame() const { return m_firedThisFrame; }
+	WeaponIndexSet GetGunsStoppedThisFrame() const { return m_stoppedThisFrame; }
+
+private:
+
+	// Handle checking and firing a given gun.
+	// Note that this currently does not nicely handle spawning multiple projectiles per timestep - i.e. timewarp or a weapon RPM higher than 3600
+	// Projectile spawns are also "snapped" to the start of a timestep if they are not direct multiples of the timestep duration
+	void Fire(WeaponState *weapon, GroupState *group);
+
+	// Calculate the position a given gun should aim at to hit the current target body
+	// This is effectively the position of the target at T+n
+	void CalcWeaponLead(WeaponState *state, vector3d position, vector3d relativeVelocity);
```

Comment:
The `vector3d` parameters should be const refs, not value.

Reply:
There is some design intent that's missing here.

`WeaponState` is intended to be as close to a POD object as possible, rather than a traditional OO "data-with-methods" class. To that end, I don't want to add any methods to it, and rely upon field initializers rather than a constructor. `CalcWeaponLead` and `GetMountTransform` are internal implementation details of the GunManager, rather than public API of the "WeaponState" object.

I'd push back slightly on the `vector3d` parameters - the final version of this function has a 122-character function signature and is only ever passed temporary variables, so I don't see a significant win here - in fact, the codegen might be worse with pass-by-reference parameters, as accesses to those parameters have to pointer-chase rather than being directly accessed from the stack. This is admittedly post-facto justification for keeping the function signature the same.

I agree with you that it is better to pass the state object by reference than by pointer however, as the pointer is not intended to be nullable.

Reply:
Are the WeaponState and other structures expected to be used outside of GunManager?  Looks to me like they should be internal state only.. in which case moving some of the functionality into the PODs makes sense to me to reduce the ultra-tight coupling.  (I see it's currently directly accessed in Ship.cpp, but marked as "abominable" and furthermore, used for introspection; it might be better to move the weapon sound playback into the GunManager, or expose relevant accessors to the GunManager to provide the required high-level information back to the Ship class).

But maybe there are future plans for more functionality around this data and it's not desired or practical to wrap all of this data into accessors; that's something I can't know.

And yes, if `CalcWeaponLead` is only ever called with temporaries, calling-by-value is actually better.. but this is very difficult to know as a reviewer reviewing a function in isolation without a comprehensive design document outlining these decisions..

Reply:
Note that the design intent regarding these specific data structs is something akin to C-with-namespaces, rather than "modern C++" - state objects do not typically mutate themselves, and have no "agency" of their own. They are instead acted upon by other code external to them, and at most might have helpers to ease setting multiple variables or accessors as needed to compute derived variables which are not stored directly.

Reply:
In general, mixing development styles in the same project should be highly discouraged.

I've nothing against a C-with-namespaces approach, but then that should be universally applied. IMHO.

Do we have an overarching style guide which future code should attempt to adhere to? This would make it easier for both contributors as well as reviewers.

Reply:
> In general, mixing development styles in the same project should be highly discouraged.
> I've nothing against a C-with-namespaces approach, but then that should be universally applied. IMHO.

I agree with this. In fact, my intent is not to use completely different development styles in different parts of the code, but rather to use different facets of the same development style where they are needed and where they make sense.

Perhaps the "C-with-namespaces" is a red herring - the important part I was attempting to get at is the distinction between data being acted upon (mutated) by an external authority, and an operation intrinsic to the data itself. An operation in the former category is extrinsic and IMO a better fit to be located in an external actor (i.e. class), while an operation in the latter category (e.g. taking the dot product of a vector) is better located in the data itself.

> Do we have an overarching style guide which future code should attempt to adhere to? This would make it easier for both contributors as well as reviewers.

Unfortunately no. We have a bit of a franken-style both at the high level and at the individual code syntax level. The closest I can point at is the general philosophy of data-oriented design, with some of the "only use as much C++ as you need" mentality of Orthodox C++.

Reply:
Thank you for your patience answering my variously rambling and discombobulated thoughts.